src/System.Linq.Expressions/src/System/Linq/Expressions/Compiler/ILGen.cs

```diff
@@ -906,45 +907,48 @@ private static void EmitNullableToNullableConversion(this ILGenerator il, Type t
             il.Emit(OpCodes.Initobj, typeTo);
             il.MarkLabel(labEnd);
             il.Emit(OpCodes.Ldloc, locTo);
+            locals.FreeLocal(locTo);
         }
 
 
-        private static void EmitNonNullableToNullableConversion(this ILGenerator il, Type typeFrom, Type typeTo, bool isChecked)
+        private static void EmitNonNullableToNullableConversion(this ILGenerator il, Type typeFrom, Type typeTo, bool isChecked, ILocalCache locals)
         {
             Debug.Assert(!typeFrom.IsNullableType());
             Debug.Assert(typeTo.IsNullableType());
-            LocalBuilder locTo = il.DeclareLocal(typeTo);
+            LocalBuilder locTo = locals.GetLocal(typeTo);
             Type nnTypeTo = typeTo.GetNonNullableType();
-            il.EmitConvertToType(typeFrom, nnTypeTo, isChecked);
+            il.EmitConvertToType(typeFrom, nnTypeTo, isChecked, locals);
             ConstructorInfo ci = typeTo.GetConstructor(new Type[] { nnTypeTo });
             il.Emit(OpCodes.Newobj, ci);
             il.Emit(OpCodes.Stloc, locTo);
             il.Emit(OpCodes.Ldloc, locTo);
+            locals.FreeLocal(locTo);
         }
 
 
-        private static void EmitNullableToNonNullableConversion(this ILGenerator il, Type typeFrom, Type typeTo, bool isChecked)
+        private static void EmitNullableToNonNullableConversion(this ILGenerator il, Type typeFrom, Type typeTo, bool isChecked, ILocalCache locals)
         {
             Debug.Assert(typeFrom.IsNullableType());
             Debug.Assert(!typeTo.IsNullableType());
             if (typeTo.GetTypeInfo().IsValueType)
-                il.EmitNullableToNonNullableStructConversion(typeFrom, typeTo, isChecked);
+                il.EmitNullableToNonNullableStructConversion(typeFrom, typeTo, isChecked, locals);
             else
                 il.EmitNullableToReferenceConversion(typeFrom);
         }
 
 
-        private static void EmitNullableToNonNullableStructConversion(this ILGenerator il, Type typeFrom, Type typeTo, bool isChecked)
+        private static void EmitNullableToNonNullableStructConversion(this ILGenerator il, Type typeFrom, Type typeTo, bool isChecked, ILocalCache locals)
         {
             Debug.Assert(typeFrom.IsNullableType());
             Debug.Assert(!typeTo.IsNullableType());
             Debug.Assert(typeTo.GetTypeInfo().IsValueType);
-            LocalBuilder locFrom = il.DeclareLocal(typeFrom);
+            LocalBuilder locFrom = locals.GetLocal(typeFrom);
             il.Emit(OpCodes.Stloc, locFrom);
             il.Emit(OpCodes.Ldloca, locFrom);
+            locals.FreeLocal(locFrom);
```

Comment:
actually, since `locals` is not passed to `EmitGetValue`, it seems ok.

---
In reply to: [100604483](https://github.com/dotnet/corefx/pull/15946#discussion_r100604483) [](ancestors = 100604483)